src/OrchardCore.Modules/OrchardCore.Contents/Sitemaps/ContentTypesSitemapSourceBuilder.cs

```diff
@@ -14,51 +16,100 @@ public class ContentTypesSitemapSourceBuilder : SitemapSourceBuilderBase<Content
 
     private readonly IRouteableContentTypeCoordinator _routeableContentTypeCoordinator;
     private readonly IContentManager _contentManager;
+    private readonly YesSqlOptions _yesSqlOptions;
     private readonly IContentItemsQueryProvider _contentItemsQueryProvider;
     private readonly IEnumerable<ISitemapContentItemExtendedMetadataProvider> _sitemapContentItemExtendedMetadataProviders;
 
     public ContentTypesSitemapSourceBuilder(
         IRouteableContentTypeCoordinator routeableContentTypeCoordinator,
         IContentManager contentManager,
+        IOptions<YesSqlOptions> yesSqlOptions,
         IContentItemsQueryProvider contentItemsQueryProvider,
-        IEnumerable<ISitemapContentItemExtendedMetadataProvider> sitemapContentItemExtendedMetadataProviders
-        )
+        IEnumerable<ISitemapContentItemExtendedMetadataProvider> sitemapContentItemExtendedMetadataProviders)
     {
         _routeableContentTypeCoordinator = routeableContentTypeCoordinator;
         _contentManager = contentManager;
+        _yesSqlOptions = yesSqlOptions.Value;
         _contentItemsQueryProvider = contentItemsQueryProvider;
         _sitemapContentItemExtendedMetadataProviders = sitemapContentItemExtendedMetadataProviders;
     }
 
     public override async Task BuildSourceAsync(ContentTypesSitemapSource source, SitemapBuilderContext context)
     {
-        var queryContext = new ContentItemsQueryContext();
-        await _contentItemsQueryProvider.GetContentItemsAsync(source, queryContext);
-
         foreach (var sciemp in _sitemapContentItemExtendedMetadataProviders)
         {
             context.Response.ResponseElement.Add(sciemp.GetExtendedAttribute);
         }
 
-        foreach (var contentItem in queryContext.ContentItems)
+        var queryContext = new ContentItemsQueryContext()
+        {
+            Take = _yesSqlOptions.CommandsPageSize,
+        };
+
+        var maxAllowed = int.MaxValue;
+
+        if (source.LimitedContentType != null)
+        {
+            queryContext.Skip = source.LimitedContentType.Skip;
+
+            if (source.LimitedContentType.Take > 0)
+            {
+                maxAllowed = source.LimitedContentType.Take;
+            }
+        }
+
+        var total = 0;
+        var isLastBatch = false;
+
+        while (true)
```

Comment:
This is not solving the issue. The content items will still be referenced by the session and kept in memory. It's only solved if each query is a different session instance, of if the objects are Detached after the query (preferred option, inside each GetContentItemsAsync call). I believe you added what it necessary to yessql to be able to do that too at the query level (or in some other way).

Reply:
Yes. Good point. We should use a read only session. I'll update that

Reply:
now I am using session with no tracking.